dask/array/routines.py

```diff
@@ -590,6 +625,15 @@ def unique(ar, return_index=False, return_inverse=False, return_counts=False):
     )
     out._chunks = tuple((np.nan,) * len(c) for c in out.chunks)
 
+    # Take the results from the unique chunks and do the following.
+    #
+    # 1. Collect all results as arguments.
+    # 2. Concatenate each result into one big array.
+    # 3. Pass all results as arguments to the internal unique again.
+    #
+    # TODO: This should be replaced with a tree reduction using this strategy.
+    # xref: https://github.com/dask/dask/issues/2851
```

Comment:
Added a TODO here in the comment to note the work that is still needed for tree reduction. More details can be found via the link to issue ( https://github.com/dask/dask/issues/2851 ).